src/modem/at.c

```diff
@@ -0,0 +1,414 @@
+        for (size_t i = 0; i < ati->urc_list.count; ++i) {
+                struct at_urc *urc = ati->urc_list.urcs + i;
+                if (0 == strncmp(msg, urc->pfx, urc->pfx_len))
+                        return urc;
+        }
+
+        return NULL;
+}
+
+static bool is_rsp_status(enum at_rsp_status *status, const char *msg)
+{
+        for (size_t i = 0; i < ARRAY_LEN(at_status_msgs); ++i) {
+                const struct at_rsp_status_msgs *sms = at_status_msgs + i;
+                if (0 == strcmp(msg, sms->str)) {
+                        *status = sms->status;
+                        return true;
+                }
+        }
+
+        return false;
+}
+
+static bool _process_msg_generic(struct at_info *ati,
+                                 const enum at_rx_state state,
+                                 const char *msg)
+{
+        if (AT_RSP_MAX_MSGS <= ati->rsp.msg_count) {
+                pr_error("[at] BUG: Received more messages than can "
+                         "handle. Dropping.");
+                return false;
+        }
+
+        ati->rx_state = state;
+        ati->rsp.msgs[ati->rsp.msg_count] = msg;
+        ++ati->rsp.msg_count;
+
+        return true;
+}
+
+static void process_urc_msg(struct at_info *ati, char *msg)
+{
+        _process_msg_generic(ati, AT_RX_STATE_URC, msg);
+
+        const bool no_status = ati->urc_ip->flags & AT_URC_FLAGS_NO_RSP_STATUS;
+        enum at_rsp_status status = AT_RSP_STATUS_NONE;
+        if (no_status || is_rsp_status(&status, msg))
+                complete_urc(ati, status);
+}
+
+static void process_cmd_msg(struct at_info *ati, char *msg)
+{
+        _process_msg_generic(ati, AT_RX_STATE_CMD, msg);
+
+        enum at_rsp_status status;
+        if (is_rsp_status(&status, msg))
+                complete_cmd(ati, status);
+}
+
+static void begin_urc_msg(struct at_info *ati, struct at_urc *urc)
+{
+        ati->urc_ip = urc;
+        ati->timing.urc_start_ms = getUptime();
+}
+
+static void at_task_run_bytes_read(struct at_info *ati, char *msg)
+{
+        /*
+         * If here, then we have read a message.  Now we have to process it.
+         * Our rx state will dictate how we process this message beacuse that
+         * allows us to know previous messages and what message type to expect.
+         */
+        switch (ati->rx_state) {
+        case AT_RX_STATE_CMD:
+                /* We are in the middle of receiving a command message. */
+                process_cmd_msg(ati, msg);
+        case AT_RX_STATE_URC:
+                /* We are in the middle of receiving a unexpected message. */
+                process_urc_msg(ati, msg);
+        case AT_RX_STATE_READY:
+                /*
+                 * We are starting a new message series.  Gotta figure out what
+                 * type of message this is before we process it.  Then we process
+                 * it appropriately.
+                 */
+                ; /* C is stupid after labels.  Workaround :\ */
```

Comment:
Good to know.  I'll give it a shot and see if it helps.  Seems like a brain dead C ism IMHO... but what are you going to do.